qemu/tests/qemu_guest_agent.py

```diff
@@ -698,38 +666,54 @@ def find_interface_by_name(interface_list, target_interface):
             error.TestFail("the interface list info after interface was down"
                            " was not as expected")
 
-    @error.context_aware
     def _action_before_fsfreeze(self, *args):
         session = self._get_session(self.params, None)
         self._open_session_list.append(session)
 
-    @error.context_aware
     def _action_after_fsfreeze(self, *args):
-        error.context("Verfiy FS is frozen in guest.")
+        logging.info("Verfiy FS is frozen in guest.")
         if not isinstance(args, tuple):
             return
 
         if not self._open_session_list:
-            raise error.TestError("Could not find any opened session")
+            test.error("Could not find any opened session")
         # Use the last opened session to send cmd.
         session = self._open_session_list[-1]
         try:
             session.cmd(self.params["gagent_fs_test_cmd"])
         except aexpect.ShellTimeoutError:
             logging.debug("FS freeze successfully.")
         else:
-            raise error.TestFail("FS freeze failed, guest still can"
-                                 " write file")
+            test.fail("FS freeze failed, guest still can"
+                      " write file")
+
+    def gagent_check_reset_reboot_shutdown(self, test, params, env):
+
+        """
+        Execute "system_reset" and send "shutdown,reboot" command to guest agent after FS freezed
+        :param test: kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment.
+        """
+        vm = env.get_vm(params["main_vm"])
+        vm.verify_alive()
+        self.gagent.fsfreeze()
```

Comment:
I will use gagent = self.gagent if I call it so many times within the function. Just a suggestion, up to you.